### ISSUE TITLE
No Access token Issue while running Express Server

### ISSUE THREAD

**Issue by sharmmoh1983** [AzureAD/microsoft-authentication-library-for-js#4904]
### Core Library

MSAL Node (@azure/msal-node)

### Core Library Version

1.3.1

### Wrapper Library

MSAL Node Extensions (@azure/msal-node-extensions)

### Wrapper Library Version

NA

### Description

I am running Test Express Server with Redis but getting issue while accessing the graph api 
/profile endpoint is nit working but sign up is successful

Please let me know if I can use this app as in docs it is mentioned not use authorization grant flow for slow acquire clients

### Error Message

Info - SilentFlowClient:acquireCachedToken - No access token found in cache for the given properties.
ClientAuthError: network_error: Network request failed. Please check network trace to determine root cause. | Fetch client threw: Error: HTTP status code 400 | Attempted to reach: https://login.microsoftonline.com/{tenantid}/oauth2/v2.0/token

### Msal Logs

Mon, 13 Jun 2022 13:22:18 GMT] : @azure/msal-node@1.9.1 : Verbose - initializeRequestScopes called
[Mon, 13 Jun 2022 13:22:18 GMT] : [9] : @azure/msal-node@1.9.1 : Verbose - buildOauthClientConfiguration called
[Mon, 13 Jun 2022 13:22:18 GMT] : [9] : @azure/msal-node@1.9.1 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/tenantid
[Mon, 13 Jun 2022 13:22:18 GMT] : [9] : @azure/msal-node@1.9.1 : Verbose - createAuthority called
[Mon, 13 Jun 2022 13:22:18 GMT] : [9] : @azure/msal-node@1.9.1 : Verbose - Silent flow client created
[Mon, 13 Jun 2022 13:22:18 GMT] : [90] : @azure/msal-common@6.4.0 : Info - SilentFlowClient:acquireCachedToken - No access token found in cache for the given properties.
ClientAuthError: network_error: Network request failed. Please check network trace to determine root cause. | Fetch client threw: Error: HTTP status code 400 | Attempted to reach: https://login.microsoftonline.com/{tenantid}/oauth2/v2.0/token

### MSAL Configuration

```javascript
NA
```

### Relevant Code Snippets

```javascript
NA
```

### Reproduction Steps

1. Start by building the wrapper:

    ```console
        cd ExpressTestApp
        npm run build
    ```

2. Then, initialize the wrapper by providing a settings file in JSON (see: [appSettings.json](./TestApp/appSettings.json)). The file looks like the following:

    ```JSON
    {
        "credentials": {
            "clientId": "CLIENT_ID",
            "tenantId": "TENANT_ID",
            "clientSecret": "CLIENT_SECRET"
        },
        "settings": {
            "homePageRoute": "/home",
            "redirectUri": "http://localhost:4000/redirect",
            "postLogoutRedirectUri": "http://localhost:4000"
        }
    }
    ```

3. Add the web API endpoints you would like to call under **resources**:

    ```JSON
    {
        // ...
        "resources": {
            "graphAPI": {
                "callingPageRoute": "/profile",
                "endpoint": "https://graph.microsoft.com/v1.0/me",
                "scopes": ["user.read"]
            },
        }
    }
    ```

5. Install and run the sample app:

    ```bash
    cd TestApp && \
        npm install && \
        npm start
    ```

6. Open a browser at `http://localhost:4000`.

### Expected Behavior

Get the Access token from Redis cache and call MS graph API

### Identity Provider

Azure B2C Basic Policy

### Browsers Affected (Select all that apply)

Firefox

### Source

External (Customer)

**Comment by hectormmg** [comment]
Hi @sharmmoh1983 , could you please share a code snippet of how you're using MSAL APIs to acquire tokens? Thanks!

**Comment by sharmam-caremonitor** [comment]
Hi @hectormmg 

I am using the sample provided in the code . Just using local Redis for session store

https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/ExpressTestApp

```
This is working
router.get('/', (req, res, next) => res.redirect('/home'));
    router.get('/home', mainController.getHomePage);
This is working
    // authentication routes
    router.get('/signin', authProvider.signIn);
 
  

    // secure routes
 This route is not working and giving error while fetching accesstoken from cache 
    router.get('/profile', authProvider.isAuthenticated, authProvider.getToken, mainController.getProfilePage); // get token for this route to call web API
   
```

Not getting the reason why error is appearing even though I am just using the existing code

Is it possible for you to run the sample and see if the same error appears.Is it a working sample?
Also there is https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/RedisTestApp what is the difference

**Comment by sharmmoh1983** [comment]
Hi @hectormmg 

I am using the sample provided in the code . Just using local Redis for session store

https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/ExpressTestApp

```
This is working
router.get('/', (req, res, next) => res.redirect('/home'));
    router.get('/home', mainController.getHomePage);
This is working
    // authentication routes
    router.get('/signin', authProvider.signIn);
 
  

    // secure routes
 This route is not working and giving error while fetching accesstoken from cache 
    router.get('/profile', authProvider.isAuthenticated, authProvider.getToken, mainController.getProfilePage); // get token for this route to call web API
   
```

Not getting the reason why error is appearing even though I am just using the existing code

Is it possible for you to run the sample and see if the same error appears.Is it a working sample?
Also there is https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/RedisTestApp what is the difference

**Comment by derisen** [comment]
@sharmmoh1983 just to confirm, are you using the sample with B2C?

Regarding the [RedisTestApp](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/RedisTestApp) -this sample demonstrates a web app with the [distributed token caching pattern](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/docs/caching.md#performance-and-security) (using Redis).

**Comment by sharmmoh1983** [comment]
Yes with default authority

**Comment by derisen** [comment]
By default authority I assume you mean a default policy/user-flow (like signup/signin?). It would be great if you could share your appSettings here. 

Regardless, with a B2C sign-in, you won't be able to acquire an access token for Graph API (or any Microsoft APIs). Please have a look at the guide: [Working with MSAL.js and Azure AD B2C](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/working-with-b2c.md).

**Comment by sharmmoh1983** [comment]
Yes you are correct .It won’t be possible to share the app settings as it contains secret values.

so how to get the access token silently from cache in case of B2c

**Comment by sharmmoh1983** [comment]
Also AAD Authentication against a B2C Tenant can get you tokens as per the link you shared .as I am not using any custom policy

**Comment by derisen** [comment]
@sharmmoh1983 I'm unable to reproduce this with the sample -you'll need to provide your appSettings (excluding client secret of course) and confirm that the account you use to sign-in have admin privileges to be able to provide consent to Graph scopes. There are limitations to calling Graph API for a B2C tenant with delegated permissions. I would also recommend going through [this sample](https://github.com/Azure-Samples/ms-identity-b2c-javascript-nodejs-management) to learn more about them.

**Comment by sharmmoh1983** [comment]
@derisen 

Thanks a lot for your support
Please find the following file
```

{
    "credentials": {
        "clientId": "xyz",
        "tenantId": "abcd",
        "clientSecret": "xxxxxxxxxx"
    },
    "settings": {
        "homePageRoute": "/home",
        "redirectUri": "http://localhost:4000/redirect",
        "postLogoutRedirectUri": "http://localhost:4000/"
    },
    "resources": {
        "graphAPI": {
            "callingPageRoute": "/profile",
            "endpoint": "https://graph.microsoft.com/v1.0/me",
            "scopes": ["user.read"]
        }
    }
```

Also I have admin privileges for the AD account used for sign in and that is why I am able to get the access token while doing sign up.

It is when I am  trying to access protected /profile resource , I am getting the error.

Also you mentioned the sample is working at your end so it seems i can use the B2C tenant with AAD profile to get access tokens silently.

Following is my requirement

1. To integrate MS calendar in my app
   So i am using msal node B2c with default sign up (AAD)  for first time sign up and it is successful.
   Then  I want my backend to make use of refresh and access token  from cache and get the calendar events for the user 


If possible , can you share the steps you followed to make the Distributed Redis sample app working

**Comment by derisen** [comment]
@sharmmoh1983 thanks for the info, I can reproduce this now. Let me debug this and get back to you.

**Comment by derisen** [comment]
@sharmmoh1983 strangely, I couldn't reproduce it again when I tried today. I'm assuming what happens is due to incorrect cache record being retrieved from redis cache. Can you please flush your redis cache, restart the redis server and give this a try again? If that does not resolve your issue, please send me the complete msal logs (set the [msal logger](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/docs/configuration.md#logger-config-options) to verbose) and the network trace (using [Fiddler](https://www.telerik.com/fiddler/fiddler-classic)) to my email -you can find it under my GitHub profile.

Please also consider using the [RedisTestApp](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/RedisTestApp) if you are planning to work with redis and distributed caching. I'll make a note to remove caching from this sample to prevent confusion.

**Comment by sharmmoh1983** [comment]
@derisen 
I am killing the redis  server and starting again  every time node starts and I could see the redis key with home account id as key

```
27.0.0.1:6379> KEYS *
1) "sess:y-------"
2) "0a------"
127.0.0.1:6379> 
```

Logs : Removed critical values 

```


> express-test-app@1.0.0 start
> node App/app.js

Msal Node Auth Code Sample app listening on port 4000!
[Fri, 17 Jun 2022 23:50:50 GMT] : @azure/msal-node@1.10.0 : Info - getAuthCodeUrl called
[Fri, 17 Jun 2022 23:50:50 GMT] : @azure/msal-node@1.10.0 : Verbose - initializeRequestScopes called
[Fri, 17 Jun 2022 23:50:50 GMT] : [dd1378a6-17e8-4afc-aa85-32334bc10976] : @azure/msal-node@1.10.0 : Verbose - buildOauthClientConfiguration called
[Fri, 17 Jun 2022 23:50:50 GMT] : [dd1378a6-17e8-4afc-aa85-32334bc10976] : @azure/msal-node@1.10.0 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/xyz
[Fri, 17 Jun 2022 23:50:50 GMT] : [dd1378a6-17e8-4afc-aa85-32334bc10976] : @azure/msal-node@1.10.0 : Verbose - createAuthority called
[Fri, 17 Jun 2022 23:50:51 GMT] : [dd1378a6-17e8-4afc-aa85-32334bc10976] : @azure/msal-node@1.10.0 : Verbose - Auth code client created
[Fri, 17 Jun 2022 23:51:50 GMT] : @azure/msal-node@1.10.0 : Info - acquireTokenByCode called
[Fri, 17 Jun 2022 23:51:50 GMT] : @azure/msal-node@1.10.0 : Verbose - initializeRequestScopes called
[Fri, 17 Jun 2022 23:51:50 GMT] : [067f1a81-ddbc-4d86-94cb-d47ec4daec92] : @azure/msal-node@1.10.0 : Verbose - buildOauthClientConfiguration called
[Fri, 17 Jun 2022 23:51:50 GMT] : [067f1a81-ddbc-4d86-94cb-d47ec4daec92] : @azure/msal-node@1.10.0 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/xyz
[Fri, 17 Jun 2022 23:51:50 GMT] : [067f1a81-ddbc-4d86-94cb-d47ec4daec92] : @azure/msal-node@1.10.0 : Verbose - createAuthority called
[Fri, 17 Jun 2022 23:51:51 GMT] : [067f1a81-ddbc-4d86-94cb-d47ec4daec92] : @azure/msal-node@1.10.0 : Verbose - Auth code client created
[Fri, 17 Jun 2022 23:51:51 GMT] : [067f1a81-ddbc-4d86-94cb-d47ec4daec92] : @azure/msal-common@7.0.0 : Info - in acquireToken call
[Fri, 17 Jun 2022 23:51:51 GMT] : [067f1a81-ddbc-4d86-94cb-d47ec4daec92] : @azure/msal-common@7.0.0 : Verbose - Persistence enabled, calling beforeCacheAccess
[Fri, 17 Jun 2022 23:51:51 GMT] : [067f1a81-ddbc-4d86-94cb-d47ec4daec92] : @azure/msal-common@7.0.0 : Verbose - Persistence enabled, calling afterCacheAccess

Response: 
: {
  authority: 'https://login.microsoftonline.com/xyx/',
  uniqueId: '',
  tenantId: '',
  scopes: [ 'openid', 'profile', 'email' ],
  account: {
    homeAccountId: '',
    environment: 'login.windows.net',
    tenantId: '',
    username: '',
    localAccountId: '',
    name: ' ',
    idTokenClaims: {
      aud: '',
      iss: 'https://login.microsoftonline.com/xyx/v2.0',
      iat: 1655509611,
      nbf: 1655509611,
      exp: 1655513511,
      name: 'xx xx',
      oid: '0xx',
      preferred_username: 'xx',
      rh: '.',
      sub: '',
      tid: '',
      uti: '',
      ver: '2.0'
    },
    nativeAccountId: undefined
  },
  idToken: '',
  idTokenClaims: {
    aud: '',
      iss: 'https://login.microsoftonline.com/xyx/v2.0',
      iat: 1655509611,
      nbf: 1655509611,
      exp: 1655513511,
      name: 'xx xx',
      oid: '0xx',
      preferred_username: 'xx',
      rh: '.',
      sub: '',
      tid: '',
      uti: '',
      ver: '2.0'
  },
  accessToken: '',
  fromCache: false,
  expiresOn: 2022-06-18T01:01:05.000Z,
  correlationId: '',
  extExpiresOn: 2022-06-18T02:10:19.000Z,
  familyId: '',
  tokenType: 'Bearer',
  state: '',
  cloudGraphHostName: '',
  msGraphHost: '',
  code: undefined,
  fromNativeBroker: false
}
[Fri, 17 Jun 2022 23:53:19 GMT] : @azure/msal-node@1.10.0 : Info - getTokenCache called
[Fri, 17 Jun 2022 23:53:19 GMT] : @azure/msal-node@1.10.0 : Verbose - initializeRequestScopes called
[Fri, 17 Jun 2022 23:53:19 GMT] : [235458d4-32b9-404d-8558-575e136bb2f6] : @azure/msal-node@1.10.0 : Verbose - buildOauthClientConfiguration called
[Fri, 17 Jun 2022 23:53:19 GMT] : [235458d4-32b9-404d-8558-575e136bb2f6] : @azure/msal-node@1.10.0 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/xyz
[Fri, 17 Jun 2022 23:53:19 GMT] : [235458d4-32b9-404d-8558-575e136bb2f6] : @azure/msal-node@1.10.0 : Verbose - createAuthority called
[Fri, 17 Jun 2022 23:53:19 GMT] : [235458d4-32b9-404d-8558-575e136bb2f6] : @azure/msal-node@1.10.0 : Verbose - Silent flow client created
[Fri, 17 Jun 2022 23:53:19 GMT] : [235458d4-32b9-404d-8558-575e136bb2f6] : @azure/msal-common@7.0.0 : Info - SilentFlowClient:acquireCachedToken - No access token found in cache for the given properties.
ClientAuthError: network_error: Network request failed. Please check network trace to determine root cause. | Fetch client threw: Error: HTTP status code 400 | Attempted to reach: https://login.microsoftonline.com/xyz/oauth2/v2.0/token
    at ClientAuthError.AuthError [as constructor] (/Users//caremonitor/microsoft-authentication-library-for-js/samples/msal-node-samples/RedisTestApp/node_modules/@azure/msal-common/dist/index.cjs.js:493:24)
    at new ClientAuthError (/Users//caremonitor/microsoft-authentication-library-for-js/samples/msal-node-samples/RedisTestApp/node_modules/@azure/msal-common/dist/index.cjs.js:787:28)
    at Function.ClientAuthError.createNetworkError (/Users///microsoft-authentication-library-for-js/samples/msal-node-samples/RedisTestApp/node_modules/@azure/msal-common/dist/index.cjs.js:830:16)
    at NetworkManager.<anonymous> (/Users//caremonitor/microsoft-authentication-library-for-js/samples/msal-node-samples/RedisTestApp/node_modules/@azure/msal-common/dist/index.cjs.js:3371:51)
    at step (/Users//caremonitor/microsoft-authentication-library-for-js/samples/msal-node-samples/RedisTestApp/node_modules/@azure/msal-common/dist/index.cjs.js:79:23)
    at Object.throw (/Users//caremonitor/microsoft-authentication-library-for-js/samples/msal-node-samples/RedisTestApp/node_modules/@azure/msal-common/dist/index.cjs.js:60:53)
    at rejected (/Users//caremonitor/microsoft-authentication-library-for-js/samples/msal-node-samples/RedisTestApp/node_modules/@azure/msal-common/dist/index.cjs.js:51:65)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)

```

**Comment by derisen** [comment]
@sharmmoh1983 just to make sure, killing the redis-server might not be enough to clear the cache, make sure to run `FLUSHALL` on redis-cli as well.

On the other hand, I think the issue is in the `getToken` method in the sample wrapper -specifically [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/b315d95797810241224a3cf118e67a1a723adec7/samples/msal-node-samples/ExpressTestApp/src/AuthProvider.ts#L287). There, we first attempt to acquire a token silently, expect it to throw `InteractionRequiredAuthError` if it fails (which it will on first attempt), then catch and handle the error interactively via `acquireTokenbyCode`. However, the actual error thrown is `ClientAuthError` instead, thus the error handling code never runs. You could try adding `error instanceof ClientAuthError` to the **if** condition there, but this is not ideal as there could be other conditions for `ClientAuthError` that cannot be handled interactively. I don't have a good solution at the moment, please keep an eye on the [original issue here](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4878).

```TypeScript
catch (error) {
    // in case there are no cached tokens, initiate an interactive call
    if (error instanceof InteractionRequiredAuthError || error instanceof ClientAuthError) {
        // ...
    } else {
        next(error);
    }
}
```

**Comment by sharmmoh1983** [comment]
@derisen  I have  done the FLUSHALL but no imapct. So please correct my understanding on below points 

-   solution will work to get the token silently for B2c AAD 
- Current implementation for Redis App need to be done to fix the issue

**Comment by derisen** [comment]
@sharmmoh1983 apologies, let me rephrase. I think what's happening here is that the `acquireTokenSilent` call fails in the sample wrapper, but msal-node doesn't throw the `InteractionRequiredAuthError ` error as expected, so the app doesn't attempt to acquire a token interactively instead for handling the error.

A workaround for now (until the [original issue](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4878) resolves) could be to add the `error instanceof ClientAuthError` condition to the if statement [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/b315d95797810241224a3cf118e67a1a723adec7/samples/msal-node-samples/ExpressTestApp/src/AuthProvider.ts#L287) as shown below:

```Typescript
catch (error) {
    // in case there are no cached tokens, initiate an interactive call
    if (error instanceof InteractionRequiredAuthError || error instanceof ClientAuthError) {
        // ...
    } else {
        next(error);
    }
}
```

Can you please try updating the code on your end and give this a try to confirm that's the issue?

**Comment by sharmmoh1983** [comment]
@derisen i am working on it and will get back soon

**Comment by sharmmoh1983** [comment]
@derisen  This will solve the exception issue but why user has to be prompted  again to give consent as the first call itself is failing to get the profile.

**Comment by derisen** [comment]
@sharmmoh1983 are you saying the call fails even after providing consent? If not, the first call failing would be expected, as at that point there is no consent provided for this scope, since delegated graph permissions can only be provided dynamically in the case of B2C, and only by an admin user.

If you need to call graph for B2C users, I would rather recommend you to setup a B2C web API/background service to call MS Graph with its own identity to query user events (using application permissions), and then have your client app to call this web API to get the user's calendar data from graph (using delegated permissions). It would be essentially a mix of these two samples: (1) [B2C web API ](https://github.com/Azure-Samples/active-directory-b2c-javascript-nodejs-webapi)(2) [B2C background service](https://github.com/Azure-Samples/active-directory-b2c-javascript-nodejs-webapi).

**Comment by sharmmoh1983** [comment]
@derisen Thanks I will implement in the same manner

**Comment by derisen** [comment]
@sharmmoh1983 sounds good, let us know if you need anything further with that effort. I'll keep this open 'till I update the sample regarding the exception issue.

**Comment by sharmmoh1983** [comment]
@derisen Just to summarise , we need the users to give consent only first up when signup/signin happens and then keep the refresh token available in redis cache to be used in subsequent call to any ms graph api call

In current scenario , i could see the consent is happening successfully and required tokens are getting generated and getting persist in redis. It is just while making any subsequent call to graph api , app is trying to fetch the token and it is failing.

So it seems B2C flow will work and we dont have to have combinations of web api/background service  if this issue gets reolved

**Comment by derisen** [comment]
@sharmmoh1983 I can't reproduce this with the sample I'm afraid. If you can mail me the complete msal logs (including correlation Ids), a Fiddler trace, your app settings (except client secret) and the user account (except password) that has this issue we can debug this further. You can find my mail on my GitHub profile.

**Comment by sharmmoh1983** [comment]
@derisen  I have already  shared  the details in previous posts. Call is failing while retrieving the token from cache first time.
Or you let me know the exact steps that you ra following at your end which is working

**Comment by derisen** [comment]
@sharmmoh1983 I'm afraid I don't have enough information to diagnose this, which is why I asked those details. The original issue and the details provided was about the `interaction_required` error/exception not being thrown, hence the failing silent token call (which is expected in this case) did not get handled. This was fixed by catching the correct exception, and initiating an interactive token call. I'm not able to reproduce your issue afterwards. I'll be happy to help if you decide to provide more details later.
